File: SDKs/python/API_consultas.py
import requests
from decimal import Decimal
from datetime import datetime

# El "mesero" conoce exactamente en qué parte del restaurante (GitHub) están los ingredientes.
# El cliente no necesita saber esta dirección.
URL_DATOS_GITHUB = "https://LucielDOD.github.io/API-Divisas/datos.json"

def Solicitar_Divisas_Disponibles() -> dict:
    """
    Obtiene la lista de códigos de divisas disponibles leyendo el JSON remoto.
    Retorna un diccionario con la respuesta formateada.
    """
    try:
        response = requests.get(URL_DATOS_GITHUB)
        response.raise_for_status()
        data = response.json()
        
        # Extraer códigos, removiendo el sufijo '-USD' (Ej: 'EUR-USD' -> 'EUR')
        codigos = [d["codigo"].replace("-USD", "") for d in data]
        codigos = sorted(list(set(codigos)))
        
        return {
            "status": "success",
            "cantidad": len(codigos),
            "divisas": codigos,
            "fecha_consulta": datetime.now().strftime("%d-%m-%Y %H:%M:%S")
        }
    except Exception as e:
        return {
            "status": "error",
            "mensaje": f"Error al obtener divisas disponibles: {str(e)}"
        }

def Solicitar_Valor_Divisa(divisa_1: str, divisa_2: str) -> dict:
    """
    Calcula el valor de divisa_1 expresado en divisa_2 extrayendo los datos remotamente.
    Retorna un diccionario con la respuesta formateada simulando una API REST.
    """
    divisa_1 = divisa_1.upper()
    divisa_2 = divisa_2.upper()

    try:
        # Descarga los precios de github
        response = requests.get(URL_DATOS_GITHUB)
        response.raise_for_status()
        data = response.json()
        
        # Mapeamos los datos { "EUR": Decimal("1.08"), ... }
        valores_en_usd = { 
            d["codigo"].replace("-USD", ""): Decimal(str(d["valor_actual"])) 
            for d in data 
        }
        
        if divisa_1 not in valores_en_usd:
            return {"status": "error", "mensaje": f"La divisa base '{divisa_1}' no se encuentra en el origen de datos."}
        if divisa_2 not in valores_en_usd:
            return {"status": "error", "mensaje": f"La divisa objetivo '{divisa_2}' no se encuentra en el origen de datos."}
            
        valor_1 = valores_en_usd[divisa_1]
        valor_2 = valores_en_usd[divisa_2]
        
        # (Divisa 1 a USD) / (Divisa 2 a USD)
        resultado = valor_1 / valor_2
        
        return {
            "status": "success",
            "codigo": f"{divisa_1}-{divisa_2}",
            "valor": round(float(resultado), 4),
            "divisa_base": divisa_1,
            "divisa_objetivo": divisa_2,
            "fecha_consulta": datetime.now().strftime("%d-%m-%Y %H:%M:%S")
        }

    except Exception as e:
        return {
            "status": "error",
            "mensaje": f"Error al calcular divisas: {str(e)}"
        }

File: SDKs/javascript/API_consultas.js
/**
 * API_consultas.js
 * 
 * SDK en JavaScript para consultar la API de Divisas de forma nativa en la web, Node.js y React Native.
 * No requiere instalación de librerías externas.
 * Utiliza fetch() nativo y resuelve promesas con el mismo formato que el SDK de Python.
 */

const URL_DATOS_GITHUB = "https://LucielDOD.github.io/API-Divisas/datos.json";

class APIConsultas {
    
    /**
     * Obtiene la lista de códigos de divisas disponibles leyendo el JSON remoto.
     * @returns {Promise<Object>} Promesa que resuelve a un objeto con el estado y las divisas.
     */
    static async solicitarDivisasDisponibles() {
        try {
            const response = await fetch(URL_DATOS_GITHUB);
            if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
            const data = await response.json();
            
            // Extraer códigos, removiendo el sufijo '-USD' y eliminando duplicados
            const codigosRudos = data.map(d => d.codigo.replace("-USD", ""));
            const codigosUnicos = [...new Set(codigosRudos)].sort();
            
            // Obtener fecha actual en la zona horaria del cliente local
            const fechaLocal = new Intl.DateTimeFormat('es-ES', { 
                dateStyle: 'short', 
                timeStyle: 'medium' 
            }).format(new Date());

            return {
                status: "success",
                cantidad: codigosUnicos.length,
                divisas: codigosUnicos,
                fecha_consulta: fechaLocal
            };
        } catch (error) {
            return {
                status: "error",
                mensaje: `Error al obtener divisas disponibles: ${error.message}`
            };
        }
    }

    /**
     * Calcula el valor de divisaBase expresado en divisaObjetivo extrayendo los datos remotamente.
     * @param {string} divisaBase - Ej. 'USD'
     * @param {string} divisaObjetivo - Ej. 'EUR'
     * @returns {Promise<Object>} Promesa con el resultado de la conversión.
     */
    static async solicitarValorDivisa(divisaBase, divisaObjetivo) {
        divisaBase = divisaBase.toUpperCase();
        divisaObjetivo = divisaObjetivo.toUpperCase();

        try {
            const response = await fetch(URL_DATOS_GITHUB);
            if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
            const data = await response.json();
            
            // Crear el mapa de valores respecto al USD
            const valoresEnUSD = {};
            data.forEach(d => {
                const codigoLimio = d.codigo.replace("-USD", "");
                valoresEnUSD[codigoLimio] = parseFloat(d.valor_actual);
            });
            
            if (!(divisaBase in valoresEnUSD)) {
                return { status: "error", mensaje: `La divisa base '${divisaBase}' no se encuentra en el origen de datos.` };
            }
            if (!(divisaObjetivo in valoresEnUSD)) {
                return { status: "error", mensaje: `La divisa objetivo '${divisaObjetivo}' no se encuentra en el origen de datos.` };
            }
            
            const valor1 = valoresEnUSD[divisaBase];
            const valor2 = valoresEnUSD[divisaObjetivo];
            
            // (Divisa 1 a USD) / (Divisa 2 a USD)
            const resultado = valor1 / valor2;
            
            // Obtener fecha actual local
            const fechaLocal = new Intl.DateTimeFormat('es-ES', { 
                dateStyle: 'short', 
                timeStyle: 'medium' 
            }).format(new Date());

            return {
                status: "success",
                codigo: `${divisaBase}-${divisaObjetivo}`,
                valor: Number(resultado.toFixed(4)),
                divisa_base: divisaBase,
                divisa_objetivo: divisaObjetivo,
                fecha_consulta: fechaLocal
            };

        } catch (error) {
            return {
                status: "error",
                mensaje: `Error al calcular divisas: ${error.message}`
            };
        }
    }
}

// Permitir importación en Node o como modulo en el navegador
if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
    module.exports = APIConsultas;
} else {
    window.APIConsultas = APIConsultas;
}
